### DISCUSSION TITLE
Efficient and elegant iterator interface

### DISCUSSION THREAD

**Discussion by lewisxy** [capnproto/capnproto#1865] [Q&A]
Consider the following interface for an iterator. This is obviously a toy example, but consider `read` could return a block of structured data in practice.
```
interface MyAPI {
  interface MyIterator {
    valid @0 () -> (result :Bool); # test if iterator is valid
    next @1 () -> (); # advance to next state
    read @2 () -> (result :UInt64); # retrieve the output of current state
  }

  makeIterator @0 (count :UInt64) -> (iterator :MyIterator);
}
```
A simple server implementation
```c++
class MyIteratorImpl final : public MyAPI::MyIterator::Server {
public:
    MyIteratorImpl(uint64_t upto) : count_(upto) {}

    kj::Promise<void> valid(ValidContext context) {
        context.getResults().setResult(valid());
        return kj::READY_NOW;
    }

    kj::Promise<void> next(NextContext context) {
        KJ_ASSERT(valid(), "iterator not valid");
        --count_;
        return kj::READY_NOW;
    }

    kj::Promise<void> read(ReadContext context) {
        KJ_ASSERT(valid(), "iterator not valid");
        context.getResults().setResult(count_);
        return kj::READY_NOW;
    }

private:
    uint64_t count_;

    bool valid() {
        return count_ > 0;
    }
};

class MyAPIImpl final : public MyAPI::Server {
public:
    kj::Promise<void> makeIterator(MakeIteratorContext context) {
        uint64_t count = context.getParams().getCount();
        context.getResults().setIterator(kj::heap<MyIteratorImpl>(count));
        return kj::READY_NOW;
    }
};
```
```c++
// create iterator on the client
MyAPI::MyIterator::Client iterator = nullptr;
{
    auto request = apiClient.makeIteratorRequest();
    request.setCount(5);
    iterator = request.send().wait(ioContext.waitScope).getIterator();
}
```
One way to use the iterator is as the following, with the familiar `for` (or `while`) loop. However, this requires a lot of waiting and can only be done on top level function. It would be much more elegant if I can wrap the entire logic using the `kj::Promise`.
```c++
// iterate through all values and store them
std::vector<uint64_t> output;
{
    for (; iterator.validRequest().send().wait(ioContext.waitScope).getResult(); iterator.nextRequest().send().wait(ioContext.waitScope)) {
        auto res = iterator.readRequest().send().wait(ioContext.waitScope);
        output.push_back(res.getResult());
    }
}
```
So here is what I came up with, using recursion.
```c++
kj::Promise<void> consumeIteratorHelper(MyAPI::MyIterator::Client&& iterator, std::vector<uint64_t>& output) {
    return iterator.validRequest().send().then([iterator = kj::mv(iterator), &output](auto&& res) mutable -> kj::Promise<void> {
        if (res.getResult()) {
            return iterator.readRequest().send().then([&output](auto&& res2) mutable {
                output.push_back(res2.getResult());
            }).then([iterator]() mutable {
                return iterator.nextRequest().send().ignoreResult();
            }).then([iterator = kj::mv(iterator), &output]() mutable {
                return consumeIteratorHelper(kj::mv(iterator), output);
            });
        } else {
            return kj::READY_NOW;
        }
    });
}

kj::Promise<std::vector<uint64_t>> consumeIteratorAsync(MyAPI::MyIterator::Client iterator) {
    std::unique_ptr<std::vector<uint64_t>> output = std::make_unique<std::vector<uint64_t>>();
    return consumeIteratorHelper(kj::mv(iterator), *output).then([output = kj::mv(output)](){
        return kj::mv(*output);
    });
}
```
And the previous block of code on the client will be equivalent to the following line.
```c++
output = consumeIteratorAsync(iterator).wait(ioContext.waitScope);
```
However, there are still a lot to details makes such implementation inefficient and not so elegant.
1. It's a lot harder to read and reason about (comparing to the `for` loop version).
2. note that `iterator` is sometimes passed by move and sometimes passed by copy. I know `Client` is copyable, but I tried to avoid copying as much as possible, and it just feels wrong to have code likes above.
3. If the [RPC order][rpc-order] were to be respected, the recursive call to `valid` shouldn't have to wait for `next` to complete. Of course, I can change the API of iterator to achieve this, but are there ways without changing it? I found it very difficult to do using existing promises API.
4. As you might already be able to tell, this is somewhat inspired by [this example][loops] from the official documentation. I don't think I fully understand the "tail call optimization" mentioned in the documentation. I tried to make the recursive call in the last continuation. Is that all it takes? Are there patterns to build complex logics with Promises like in this iterator case? What if I want to add some error handling logic?

[rpc-order]: https://github.com/capnproto/capnproto/discussions/1862
[loops]: https://github.com/capnproto/capnproto/blob/v2/kjdoc/tour.md#loops

**Comment by kentonv** [comment]
I believe you've used recursion correctly here to form a loop. Promise tail call optimization should apply so you don't get an infinite promise chain.

If you want the code to look nicer, I'd recommend using coroutines. KJ promises support C++20 coroutines. You can take your code that does `p.wait(waitScope)` and change it to `co_await p` instead, then you don't need a WaitScope so the code can run anywhere, not just at the top level.
